vscode-lean4/src/utils/clientProvider.ts

```diff
@@ -29,6 +29,9 @@ export class LeanClientProvider implements Disposable {
     private clientRemovedEmitter = new EventEmitter<LeanClient>()
     clientRemoved = this.clientRemovedEmitter.event
 
+    private activeClientStoppedEmitter = new EventEmitter<string>()
+    activeClientStopped = this.activeClientStoppedEmitter.event
```

Comment:
```suggestion
    private clientStoppedEmitter = new EventEmitter<[LeanClient, string]>()
    clientStopped = this.clientStoppedEmitter.event
```

The extension can start more than LSP server process (= LeanClient = `lean --server`).  For example, if you open several projects in a workspace, each of these project directories will have a different server.  (This server process then in turn starts so-called worker processes, i.e., `lean --worker`, one for each file.  But the extension doesn't see that.)

So we should remember which LSP server stopped.

(I've just noticed that we already don't handle this cleanly in `Main`.  Apparently `Waiting for Lean server to start...` is only shown before the first server has started.  But not for the second one.  Ideally we'd show the message for the Lean server of the current file, of course.  We can fix that another time though.)

Reply:
I'll commit the suggestion and we can handle this properly in Main in next iterations 👍

Reply:
Actually, Gabriel is correct, I will work with Mariana on this one.

Reply:
> the client the infoview cares about

The issue here is that the infoview needs to know the status of *all* clients, not just the active one. For example, using the 'pin' feature a user may pin an `Info` component from `A.lean` while also having `B.lean` open, and these two files can communicate with different clients.

But i would agree with Gabriel that this is a larger change that should be done in another PR, so I think leaving the `activeClientStoppedEmitter` as-is is good for now. For example, iirc the `RpcSession` manager does not really handle multiple clients at the moment.

Reply:
Then that is a huge new feature we should not try and do in this PR.  but I still maintain it is redundant for object A to raise an event with parameter A, the receiver of the event knows who it came from because added the event handler. 
 For example, notice that ClientProvider has a captured local variable named "client", so it doesn't need the client again in the event args.

Reply:
> Then that is a huge new feature we should not try and do in this PR.

I agree.

> I still maintain it is redundant for object A to raise an event with parameter "this", the receiver of the event knows who it came from because it added the event handler.

Sorry, I do not understand. `activeClientStoppedEmitter` is an event on `LeanClientProvider` which stores multiple `LeanClient`s, so a `LeanClient` parameter does not seem redundant there (it is not `this : LeanClientProvider`).

Reply:
Woops you are right, I was confusing it with `LeanClient` which has `private stoppedEmitter = new EventEmitter<string>()`. And I now understand what you mean by info view showing multiple cursor locations from potentially different LeanClients because of the "pinning" feature as shown below.  I will work with Mariana on a solution that handles this properly.  Clearly `activeClientStoppedEmitter` is not sufficient.  It needs to be `clientStoppedEmitter` and the `this.webviewPanel?.api.serverStopped` probably needs to pass something about the server, for example, should the `pinned` state now say `client stopped` ?  Or should it stay alive pinned or perhaps it should move to "paused" ?  And if user tries to unpause it then it says `client stopped` ?

![image](https://user-images.githubusercontent.com/18707114/168659960-51c5a832-9a93-4897-9095-cc6dd666a6f4.png)

Reply:
We need to keep track of which server stopped even if you don't pin anything.  Instead of pinning information from another file, you could just switch back-and-forth between files.  You'd expect the "server stopped" message to update when you switch files.

> should the `pinned` state now say `client stopped` ? Or should it stay alive pinned or perhaps it should move to "paused" ? And if user tries to unpause it then it says `client stopped` ?

I think it should just say "(server stopped)" (or even shorter: "(crashed)").  Changing state shouldn't be necessary.

Reply:
Good, point, right now if I load the above multi-workspace test, and edit Foo.lean then Main.lean and then kill the worker process for -foo, then switch back to Foo.lean it does not show "server stopped unexpectedly", in fact our LeanClient doesn't even get notified of any of that because of https://github.com/leanprover/vscode-lean4/issues/175 and the --server process auto-restarts the server for Foo workspace and when you click somewhere else in the file it starts working again.  Can you think of another repro for the issue you are raising that is not #175 ?

Reply:
@lovettchris I think the repro would be to kill the `--server` (watchdog) process for one of the workspaces, not the worker.

Reply:
Ah, of course, ok, that one pops up this error message:

![image](https://user-images.githubusercontent.com/18707114/168673186-22dff51b-7520-4956-a109-9aa499f4cf76.png)

But the InfoView for Foo.lean doesn't say "Lean language server has stopped" until I edit something.  So great, this is the repro:

1. Load multi-folder workspace 
2. Edit Foo.lean from workspace "Foo"
3. Edit Main.lean from workspace "Main"
4. Kill the --server for the "Foo" workspace and do not click the "Restart Lean Language Server" button that pops up.
5. Switch to Foo.lean.

Notice now that "Infoview" does not say "Lean language server has stopped" until you "edit" the file - which is a bug.

Reply:
Resolved